### PR TITLE
Fix Inductor signbit loss in log_ndtr edge cases (#187336)

### LINKED ISSUE #187336: [Inductor] torch.compile loses signbit in torch.special.log_ndtr, causing discrepancies in sign-sensitive ops (e.g., copysign, atan2)
### 🐛 Describe the bug
When compiling `torch.special.log_ndtr` with the `inductor` backend, the output for large positive inputs (e.g., `x = 100.0`) evaluates to `0.0`, whereas eager mode evaluates to `-0.0`. 

While `-0.0` and `0.0` evaluate to `True` under `torch.allclose`, this difference in the signbit causes discrepancies in downstream operations that are sensitive to the sign of zero (such as `torch.copysign`, `torch.atan2`, etc.). For example, using `torch.copysign` on the output yields `-1.0` in eager mode, but `1.0` in compiled mode.

### Reproduce script
```python
import torch

print("torch:", torch.__version__)
print("cuda:", torch.version.cuda)

def fn(x):
    y = torch.special.log_ndtr(x)
    return torch.copysign(torch.tensor(1.0, device=x.device, dtype=x.dtype), y)

x = torch.tensor([100.0], device="cuda", dtype=torch.float32)

y_eager = fn(x)
y_compiled = torch.compile(fn, backend="inductor", fullgraph=True)(x)
    
print("eager:", y_eager)
print("compiled:", y_compiled)
print("equal:", torch.equal(y_eager, y_compiled))
print("allclose:", torch.allclose(y_eager, y_compiled))
```

### Output logs
```text
torch: 2.14.0.dev20260612+cu130
cuda: 13.0
eager: tensor([-1.], device='cuda:0')
compiled: tensor([1.], device='cuda:0')
equal: False
allclose: False
```

### Root cause observation
The discrepancy originates from `torch.special.log_ndtr(x)` itself. If we only return `torch.special.log_ndtr(x)` for the same input, the outputs are:
* **Eager:** `tensor([-0.], device='cuda:0')`
* **Compiled:** `tensor([0.], device='cuda:0')`

### Versions

PyTorch version: 2.14.0.dev20260612+cu130
CUDA used to build PyTorch: 13.0
OS: Ubuntu 24.04.4 LTS (x86_64)
Is CUDA available: True
CUDA runtime version: 12.0.140

cc @mruberry @kshitij12345 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by l1ve709**
**Title:** `[Inductor] Fix signbit loss in torch.special.log_ndtr`

## Description

Fixes #187336 
*(Credit to @xyt66665000 for discovering and reporting the issue with a minimal repro. Authored with the assistance of an AI coding agent.)*

### The Issue
When compiling `torch.special.log_ndtr` with Inductor, extremely large positive inputs incorrectly evaluate to `0.0` instead of `-0.0`. 

This occurs due to fast-math optimizations (specifically `-fno-signed-zeros`) in Inductor's C++ and Triton codegen. The core decomposition evaluates `torch.log1p(-torch.erfc(t) / 2)`. For large inputs, `erfc(t)` flushes to `0.0`, resulting in `log1p(-0.0)`. Fast-math simplifies this to `log(1.0) = 0.0`, swallowing the negative signbit. This causes silent numerical instability when the output is fed into sign-sensitive operations like `copysign`, `atan2`, or reciprocals.

### The Fix
Mathematically, the standard normal CDF is bounded by $\le 1$, meaning its logarithm `log_ndtr(x)` is strictly bounded by $\le 0$. 

This PR specifically targets Inductor by adding an override in `torch/_inductor/decomposition.py`. It excludes the core decomposition from `decomps_to_exclude` and re-registers it wrapped in `torch.copysign(res, -1.0)` to safely enforce the signbit.

**Why this approach?**
1. **Targeted Scope:** By overriding the decomposition solely in Inductor's registry, we avoid touching the universal eager `_refs`, isolating the fix to where the fast-math compiler flag actually exists.
2. **Robustness:** Protects the signbit against downstream fast-math assumptions without requiring intrusive compiler flags.
3. **Correctness:** Ensures `-0.0` is consistently preserved for large positive inputs.
4. **Gradient Safety:** Preserves `NaN` correctness and gradients safely.

## Testing
- **Edge Cases:** Tested `NaN` and explicit `-inf` edge cases, as well as branch boundaries (`x=1.0`).
- **Regression Test:** Added `test_log_ndtr_signbit` parameterized across `device=["cpu", "cuda"]` and `dtype=[float32, float64]` to ensure signbit preservation under `torch.compile(backend="inductor")`.
- **MPS skipped:** Skipped the test on Apple Metal (MPS) since MPS doesn't currently implement `aten::special_log_ndtr.out`, and `-fno-signed-zeros` is specific to C++/Triton codegen anyway.

## Pull Request Type
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by l1ve709** [comment]
There are a few things I’ve missed, I’m interested

**Comment by l1ve709** [comment]
`@jansel` addressed all the points from Claude's review. 
﻿
Instead of patching `_refs`, I added `special_log_ndtr` to `decomps_to_exclude`. This forces Inductor to use the native `calc_log_ndtr` lowering, which naturally preserves the `-0.0` signbit without fighting fast-math. 
﻿
Also moved the regression test into `test_torchinductor.py` and cleaned up the `skipTest` blocks.
﻿
Could you approve the CI workflows when you have a chance? Thanks!

**Comment by jansel** [comment]
@claude review these changes again

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -146,6 +146,7 @@
     aten._foreach_addcdiv_,
     aten.lerp,
     aten.lerp_,
+    aten.special_log_ndtr,  # native lowering preserves -0.0 signbit; decomp loses it under -fno-signed-zeros (#187336)
```
Comment (jansel) [inline torch/_inductor/decomposition.py]: Is this just causing this function to fallback?  Can we fix the underlying issue in the decomp rather than skipping it?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 1 above)
Comment (l1ve709) [reply torch/_inductor/decomposition.py]: Good call, excluding the decomp forced a FallbackKernel on GPU, and
fp16/Halide broke because those backends don't have a native log_ndtr
lowering in OverridesData.

You're right that the underlying issue should be fixed in the decomp.
Turns out the CPU signbit loss isn't really about the decomp math,  it's
that Inductor compiles the C++ lowering under -fno-signed-zeros, which
just strips the -0.0 out of log1p(-0.0).

Rather than touching the compiler flags globally, I reverted the decomp
exclusion (so fusion's back for all backends) and added
torch.copysign(res, -1.0) in the Python decomp instead. log_ndtr is always
non-positive anyway, so this pins the sign directly instead of hoping the
arithmetic survives fast-math.

Still confirming end-to-end  I'll watch the CI run across CPU/CUDA/Halide
to make sure the signbit actually holds. Tests are also updated:
@parametrize, an is_dtype_supported guard, and a mid-range -5.0 check.

**Comment by l1ve709** [comment]
@jansel The two CI failures are backend limitations, not logic bugs:
- **Halide:** no `copysign` support
- **Triton-CPU:** no `erfcx` in libdevice

Should I add skip checks for these backends in `test_log_ndtr_signbit`, or is there a preferred approach?

**Comment by jansel** [comment]
Skips for Halide/Triton-CPU are fine.

**Comment by l1ve709** [comment]
@jansel Done, added `_expected_failure_halide` and `_expected_failure_triton_cpu` flags to `test_log_ndtr_signbit` using the existing xfail_prop mechanism.

**Review by jansel — APPROVED** [review]

**Comment by l1ve709** [comment]
@jansel I guess we're done here. Could you please trigger the merge again? TY.

**Review by jansel — CHANGES_REQUESTED** [review]
Failing tests look related

**Comment by l1ve709** [comment]
@jansel Pushed a fix, moved the `copysign` wrapper from the universal decomposition (`_refs/special`) to an inductor-specific decomposition in `torch/_inductor/decomposition.py`. Since the `-fno-signed-zeros` issue only affects inductor's C++ codegen, this should isolate the fix from other backends (MPS, ROCm) that use the universal decomposition. Waiting on CI to confirm..

**Comment by l1ve709** [comment]
@jansel Pushed a quick fix for the remaining 3 macos failures. 

The test was failing with `NotImplementedError` because `aten::special_log_ndtr.out` doesn't have an MPS lowering yet. I just skipped it on MPS since this signbit issue is specific to C++/Triton codegen anyway.

Does this approach look good to you? If so, could you approve the CI run?

**Review by jansel — APPROVED** [review]
